File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        answer = [0]*n
        stack = [] # (index, temperature)
        for i, t in enumerate(temperatures):
            while stack and stack[-1][1] < t:
                topI, _ = stack.pop()
                answer[topI] = i - topI
            stack.append((i, t))
        return answer
                    

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution{

	public int[] dailyTemperatures(int[] temperatures){
	// [1,2,4,1,2,4,3]
	// asnwer 
	Stack<Pair<Integer, Integer>> st = new Stack<>();
	int[] answer = new int[temperatures.length];
	
	for(int i = temperatures.length - 1; i >=0; --i){
		
		while(!st.isEmpty() && temperatures[i] >= st.peek().getKey()){
			
			st.pop();
}

answer[i] = st.isEmpty() == true ? 0: st.peek().getValue() - i;
st.push(new Pair<Integer, Integer>(temperatures[i], i));
}

return answer;
}}
